src/coordinator/coordinator.hpp

```diff
@@ -0,0 +1,259 @@
+
+using CoordinatorReadRequests = std::variant<GetShardMapRequest, HeartbeatRequest>;
+using CoordinatorReadResponses = std::variant<GetShardMapResponse, HeartbeatResponse>;
+
+class Coordinator {
+  ShardMap shard_map_;
+  uint64_t highest_allocated_timestamp_;
+
+  /// Query engines need to periodically request batches of unique edge IDs.
+  uint64_t highest_allocated_edge_id_;
+
+  CoordinatorReadResponses static HandleRead(HeartbeatRequest && /* heartbeat_request */) {
+    return HeartbeatResponse{};
+  }
+
+  CoordinatorReadResponses HandleRead(GetShardMapRequest && /* get_shard_map_request */) {
+    GetShardMapResponse res;
+    res.shard_map = shard_map_;
+    return res;
+  }
+
+  CoordinatorWriteResponses ApplyWrite(HlcRequest &&hlc_request) {
+    HlcResponse res{};
+
+    auto hlc_shard_map = shard_map_.GetHlc();
+
+    MG_ASSERT(!(hlc_request.last_shard_map_version.logical_id > hlc_shard_map.logical_id));
+
+    res.new_hlc = Hlc{
+        .logical_id = ++highest_allocated_timestamp_,
+        // TODO(tyler) probably pass some more context to the Coordinator here
+        // so that we can use our wall clock and enforce monotonicity.
+        // .coordinator_wall_clock = io_.Now(),
+    };
+
+    // Allways return fresher shard_map for now.
+    res.fresher_shard_map = std::make_optional(shard_map_);
+
+    return res;
+  }
+
+  CoordinatorWriteResponses ApplyWrite(AllocateEdgeIdBatchRequest &&ahr) {
+    AllocateEdgeIdBatchResponse res{};
+
+    uint64_t low = highest_allocated_edge_id_;
+
+    highest_allocated_edge_id_ += ahr.batch_size;
+
+    uint64_t high = highest_allocated_edge_id_;
+
+    res.low = low;
+    res.high = high;
+
+    return res;
+  }
+
+  /// This splits the shard immediately beneath the provided
+  /// split key, keeping the assigned peers identical for now,
+  /// but letting them be gradually migrated over time.
+  CoordinatorWriteResponses ApplyWrite(SplitShardRequest &&split_shard_request) {
+    SplitShardResponse res{};
+
+    if (split_shard_request.previous_shard_map_version != shard_map_.shard_map_version) {
+      res.success = false;
+    } else {
+      res.success = shard_map_.SplitShard(split_shard_request.previous_shard_map_version, split_shard_request.label_id,
+                                          split_shard_request.split_key);
+    }
+
+    return res;
+  }
+
+  /// This adds the provided storage engine to the standby storage engine pool,
+  /// which can be used to rebalance storage over time.
+  CoordinatorWriteResponses static ApplyWrite(RegisterStorageEngineRequest && /* register_storage_engine_request */) {
+    RegisterStorageEngineResponse res{};
+    // TODO
+
+    return res;
+  }
+
+  /// This begins the process of draining the provided storage engine from all raft
+  /// clusters that it might be participating in.
+  CoordinatorWriteResponses static ApplyWrite(DeregisterStorageEngineRequest && /* register_storage_engine_request */) {
```

Comment:
west vs east static :D Plz move all static at the front to be consistent with the codebase

Reply:
done